### PR TITLE
second order reference model filter with optional rate feedback

### PR DESCRIPTION
**Pull request by tstastny**
**Describe problem solved by this pull request**
Reference models can be used as filters which exhibit a particular, chosen (reference) dynamic behavior. This PR implements a simple second order transfer function which can be used as such a reference model, additionally with rate feedback. The system is parameterized by explicitly set natural frequency and damping ratio. Another nice feature is that the output state and rate are kinematically consistent.

This PR doesn't use this filter in an application just yet.. but there are some future PR's in the pipeline which will benefit from this filter formulation. Example applications: Input filter for attitude / rate commands before reaching the attitude / rate controllers, respectively. Smoothing out airspeed estimates with aid from accelerometer measurements (I believe this is even happening in its own implementation in TECS currently).

Block diagram:
second-order-reference-model drawio

**Describe possible alternatives**
Why not just use a regular delayed second order filter? Using the rate feedback aids tracking if it is available. If one does not have such feedback, the filter will still work as a 2nd order system with zero-rate setpoint.

A method within the class which could be extended / improved is the integration function. ATM is a simple forward Euler step, assuming whoever uses this filter runs it fast enough for this to be ok.

**Test data / coverage**
As can be seen from the plots below, the filter running without rate feedback has noticeable delay w.r.t. the raw (noisier) inputs. These raw data here was generated by integrating some random walk accelerations twice.
vel_plot_2nd_o_filt
pos_plot_2nd_o_filt

Unit tests are also included in this PR. If there are some other specific unit test cases to add, please let me know. Further open to any alterations for generality (e.g. continuous -> discrete time considerations, naming, additional parameter inputs, etc).

TODO:
- [x] convert to discrete (bilinear transform) formulation

### CODE REVIEW

**Comment by tstastny** [comment]
Forgot to also mention #18026 will probably use this soon for some output filtering.

Another note - seems omnibus f4 is out of flash. Not sure what do do about this :S

**Comment by tstastny** [comment]
@bresch @dagar  there's also no time step handling in this class at the moment. so the double integrator could do some big jumps if the filter isn't called frequently enough. -- how do you regularly prefer to handle this?
1. user of the class' responsibility?
2. min max timestep caps?
3. better integrator? (would involve e.g. more intermediate integration steps.. but this also bumps computation time)

**Comment by tstastny** [comment]
@bresch I added the fixes/additions we talked about yesterday.

> user of the class' responsibility?

Note added to warn filter performance is *not* agnostic to sample time.

> min max timestep caps? (+ resets)

Done.

> better integrator?

Not for now.

If I have some more headspace to write out the bilinear transform and think how to represent in MIMO form later this week I'll make a separate PR. Otherwise this would probably be ok from my end to merge if all ok to you.

File: src/lib/mathlib/math/filter/second_order_reference_model.hpp [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+	// A conservative multiplier on sample frequency to bound the maximum time step
+	static constexpr float SAMPLE_RATE_MULTIPLIER = 10.0f;
+
+	// (effective, no mass) Spring constant for second order system [s^-2]
+	float spring_constant_{0.0f};
+
+	// (effective, no mass) Damping coefficient for second order system [s^-1]
+	float damping_coefficient_{0.0f};
+
+	T filter_state_{}; // [units]
```
Comment (bresch) [inline src/lib/mathlib/math/filter/second_order_reference_model.hpp:176]:
```suggestion
	T _filter_state{}; // [units]
```
Not really important, but we're usually **prefixing** with underscore for member variables

**Review by bresch — DISMISSED** [review]
I verified the equations, looks good to me

File: src/lib/mathlib/math/filter/second_order_reference_model.hpp (hunk 1 above)
Comment (tstastny) [reply src/lib/mathlib/math/filter/second_order_reference_model.hpp:176]: ill wait until our code style discussion/decision finalizes to change this :)

**Comment by tstastny** [comment]
> I verified the equations, looks good to me

thanks @bresch - I converted back to draft now to implement the discrete formulation, once that's done i'll tag you again :D

File: src/lib/mathlib/math/test/second_order_reference_model_test.cpp [hunk 2]
```diff
@@ -0,0 +1,132 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2019 ECL Development Team. All rights reserved.
```
Comment (bresch) [inline src/lib/mathlib/math/test/second_order_reference_model_test.cpp:3]:
```suggestion
 *   Copyright (c) 2022 ECL Development Team. All rights reserved.
```

**Review by bresch — APPROVED** [review]
I checked compared the equations with the python derivation. Looks good

**Comment by tstastny** [comment]
thanks @bresch !

**Pull request merged**